File: spring-data-requery-test/src/main/java/org/springframework/data/requery/domain/upsert/AbstractUpsertLocation.java
package org.springframework.data.requery.domain.upsert;

import io.requery.*;
import lombok.Getter;
import org.springframework.data.requery.domain.AbstractPersistable;

import java.util.Objects;

/**
 * @author Diego on 2018. 6. 10..
 */
@Getter
@Entity(name = "UpsertLocation")
@Table(name = "upsert_location")
public abstract class AbstractUpsertLocation extends AbstractPersistable<Integer> {

    @Key
    @Generated
    @Column(name = "locationId")
    protected Integer id;

    @Column
    protected String name;

    @Embedded
    public AbstractUpsertAddress address;

    @Override
    public int hashCode() {
        return Objects.hash(address);
    }
}

File: spring-data-requery-kotlin/src/test/kotlin/com/coupang/springframework/data/requery/domain/upsert/AbstractUpsertLocation.kt
package com.coupang.springframework.data.requery.domain.upsert

import com.coupang.kotlinx.core.hashOf
import com.coupang.kotlinx.objectx.uninitialized
import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*

/**
 * AbstractUpsertLocation
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
@Entity
@Table(name = "UpsertLocation")
abstract class AbstractUpsertLocation: AbstractPersistable<Int>() {

    @get:Key
    @get:Generated
    @get:Column(name = "locationId")
    abstract override val id: Int?

    @get:Column
    abstract var name: String?

    // NOTE: Embedded 에 해당하는 놈은 abstract 를 빼야 한다.
    @get:Embedded
    var address: AbstractUpsertAddress = uninitialized()

    override fun hashCode(): Int {
        return hashOf(address)
    }

}